File: Java-Kotlin/src/main/java/jianzhi/jz3.java
package jianzhi;

import common.datastructure.ListNode;

import java.util.ArrayList;
import java.util.Stack;

public class jz3 {

    public ArrayList<Integer> printListFromTailToHead(ListNode listNode) {
        if (listNode == null) {
            return new ArrayList<>();
        }

        Stack<Integer> stack = new Stack<>();
        ListNode runner = listNode;
        while (runner != null) {
            stack.push(runner.val);
            runner = runner.next;
        }

        ArrayList<Integer> result = new ArrayList<>();
        while (!stack.isEmpty()) {
            result.add(stack.pop());
        }
        return result;
    }

    public ArrayList<Integer> printListFromTailToHead_2(ListNode listNode) {
        if (listNode == null) {
            return new ArrayList<>();
        }

        ListNode newHead = reverse(listNode);
        ArrayList<Integer> result = new ArrayList<>();
        while (newHead != null) {
            result.add(newHead.val);
            newHead = newHead.next;
        }
        return result;
    }

    private ListNode reverse(ListNode listNode) {
        ListNode newHead = null;
        ListNode next = null;
        ListNode runner = listNode;
        while (runner != null) {
            next = runner.next;
            runner.next = newHead;
            newHead = runner;
            runner = next;
        }
        return newHead;
    }

}

File: Go-Solutions/nowcoder/jz3.go
package nowcoder

func printListFromTailToHead(head *ListNode) []int {
	// write code here

	if head == nil {
		return []int{}
	}

	stack := make([]int, 0)
	runner := head
	for runner != nil {
		stack = append(stack, runner.Val)
		runner = runner.Next
	}

	res := make([]int, 0)
	for len(stack) != 0 {
		value := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		res = append(res, value)
	}
	return res
}

func printListFromTailToHead_2(head *ListNode) []int {
	if head == nil {
		return []int{}
	}

	newHead := reversejz3(head)
	res := make([]int, 0)
	for newHead != nil {
		res = append(res, newHead.Val)
		newHead = newHead.Next
	}

	return res
}

func reversejz3(head *ListNode) *ListNode {
	var newHead *ListNode = nil
	var next *ListNode = nil
	runner := head
	for runner != nil {
		next = runner.Next
		runner.Next = newHead
		newHead = runner
		runner = next
	}
	return newHead
}
